src/message.rs

```diff
@@ -0,0 +1,211 @@
+
+            match input.get(..1) {
+                // If a param started with a :, that means the rest of the input
+                // is a single trailing param.
+                Some(":") => {
+                    params.push(&input[1..]);
+                    break;
+                }
+
+                // Anything else is a normal param.
+                Some(_) => {
+                    match input.find(' ') {
+                        Some(loc) => {
+                            params.push(&input[..loc]);
+                            // Update input to point to everything after the space
+                            input = &input[loc..];
+                        }
+
+                        // If we couldn't find a space, the rest of the string
+                        // is the param.
+                        None => {
+                            params.push(input);
+                            break;
+                        }
+                    }
+                }
+
+                // If we didn't get anything, this means we're at the end of the
+                // string.
+                None => {
+                    break;
+                }
+            }
+        }
+
+        if params.is_empty() {
+            return Err(ParseError::CommandError("command not found".to_string()));
+        }
+
+        // Take the first param as the command. Note that we've already checked
+        // if params is empty, so we can unwrap this safely.
+        let (command, params) = params.split_first().unwrap();
+
+        Ok(Message {
+            tags,
+            prefix,
+            command,
+            params: params.to_vec(),
+        })
+    }
+}
+
+impl<'a> fmt::Display for Message<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if !self.tags.is_empty() {
+            f.write_char('@')?;
+
+            for (i, (k, v)) in self.tags.iter().enumerate() {
+                // We need to insert a separator for everything other than the
+                // first value.
+                if i != 0 {
+                    f.write_char(';')?;
+                }
+
+                f.write_str(k)?;
+                if !v.is_empty() {
+                    f.write_char('=')?;
+                }
+
+                for c in v.chars() {
+                    match escape_char(c) {
+                        Some(escaped_str) => f.write_str(escaped_str)?,
+                        None => f.write_char(c)?,
+                    }
+                }
+            }
+
+            f.write_char(' ')?;
+        }
+
+        if let Some(prefix) = &self.prefix {
+            f.write_str(":")?;
+            f.write_str(prefix)?;
+            f.write_char(' ')?;
+        }
+
+        f.write_str(&self.command)?;
+
+        if let Some((last, params)) = self.params.split_last() {
+            for param in params {
+                f.write_char(' ')?;
+                f.write_str(param)?;
```

Comment:
I could see this either way. I personally think the current implementation is clearer... since the only reason flipping this would work is because I always write out the last param as a trailing param (starting with a `:`). Changing this would also require me to add a write_char after writing the command.

I'll think on it.

Reply:
Whoopsies, disregard this comment.  I sorta skimmed over the fact that you'll always be writing a command.  Sorry!

Reply:
Got it!